thirdparty/Qt-Color-Widgets/src/color_wheel.cpp

```diff
@@ -0,0 +1,716 @@
+    QWidget(parent), p(new Private(this))
+{
+    setDisplayFlags(FLAGS_DEFAULT);
+    setAcceptDrops(true);
+    connect(this, SIGNAL(colorChanged(QColor)), this, SIGNAL(harmonyChanged()));
+}
+
+ColorWheel::~ColorWheel()
+{
+    delete p;
+}
+
+QColor ColorWheel::color() const
+{
+    return p->color_from(p->hue, p->sat, p->val, p->alpha);
+}
+
+QList<QColor> ColorWheel::harmonyColors() const
+{
+    QList<QColor> result;
+    result.push_back(color());
+    for (auto const& harmony : p->ring_editors)
+    {
+        auto hue = normalize(p->hue+harmony.hue_diff);
+        result.push_back(p->color_from(hue, p->sat, p->val, 1));
+    }
+    return result;
+}
+
+unsigned int ColorWheel::harmonyCount() const
+{
+    return 1 + p->ring_editors.size();
+}
+
+QSize ColorWheel::sizeHint() const
+{
+    return QSize(p->wheel_width*5, p->wheel_width*5);
+}
+
+qreal ColorWheel::hue() const
+{
+    if ( (p->display_flags & COLOR_LCH) && p->sat > 0.01 )
+        return color().hueF();
+    return p->hue;
+}
+
+qreal ColorWheel::saturation() const
+{
+    return color().hsvSaturationF();
+}
+
+qreal ColorWheel::value() const
+{
+    return color().valueF();
+}
+
+unsigned int ColorWheel::wheelWidth() const
+{
+    return p->wheel_width;
+}
+
+void ColorWheel::setWheelWidth(unsigned int w)
+{
+    p->wheel_width = w;
+    p->render_inner_selector();
+    update();
+}
+
+void ColorWheel::paintEvent(QPaintEvent * )
+{
+    QPainter painter(this);
+    painter.setRenderHint(QPainter::Antialiasing);
+    painter.translate(geometry().width()/2,geometry().height()/2);
+
+    // hue wheel
+    if(p->hue_ring.isNull())
+        p->render_ring();
+
+    painter.drawPixmap(-p->outer_radius(), -p->outer_radius(), p->hue_ring);
+
+    // hue selector
+    p->draw_ring_editor(p->hue, painter, Qt::black);
+
+    for (auto const& editor : p->ring_editors)
+    {
+        auto hue = p->hue+editor.hue_diff;
+        // TODO: better color for uneditable indicator
+        auto color = editor.editable ? Qt::white : Qt::gray;
+        p->draw_ring_editor(hue, painter, color);
+    }
+
+    // lum-sat square
+    if(p->inner_selector.isNull())
+        p->render_inner_selector();
+
+    painter.rotate(p->selector_image_angle());
+    painter.translate(p->selector_image_offset());
+
+    QPointF selector_position;
+    if ( p->display_flags & SHAPE_SQUARE )
+    {
+        qreal side = p->square_size();
+        selector_position = QPointF((1.0-p->sat)*side, p->val*side);
+    }
+    else if ( p->display_flags & SHAPE_TRIANGLE )
+    {
+        qreal side = p->triangle_side();
+        qreal height = p->triangle_height();
+        qreal slice_h = side * p->val;
+        qreal ymin = side/2-slice_h/2;
+
+        selector_position = QPointF(p->val*height, ymin + p->sat*slice_h);
+        QPolygonF triangle;
+        triangle.append(QPointF(0,side/2));
+        triangle.append(QPointF(height,0));
+        triangle.append(QPointF(height,side));
+        QPainterPath clip;
```

Comment:
When compiling the osx version for OTX using Qt 5.15, I get the following compile time error:
```
/Users/travis/build/manongjohn/OTX/thirdparty/Qt-Color-Widgets/src/color_wheel.cpp:412:22: error: variable has incomplete type 'QPainterPath'
        QPainterPath clip;
                     ^
/usr/local/opt/qt/lib/QtGui.framework/Headers/qmatrix.h:54:7: note: forward declaration of 'QPainterPath'
class QPainterPath;
      ^
```
Can easily be fixed by adding `#include <QPainterPath>` at the top of this file and still compiles for prior versions.